symengine/sets.cpp

```diff
@@ -0,0 +1,211 @@
+                    ((this->left_open_ == s.left_open_) and (this->right_open_ == s.right_open_)));
+        }
+        return false;
+    }
+
+    int Interval::compare(const Basic &o) const {
+        SYMENGINE_ASSERT(is_a<Interval>(o))
+        const Interval &s = static_cast<const Interval &>(o);
+        if ((eq(*this->start_, *s.start_) and eq(*this->end_, *s.end_)) and
+                    ((this->left_open_ == s.left_open_) and (this->right_open_ == s.right_open_)))
+            return 0;
+        else
+            throw std::runtime_error("Not implemented");
+    }
+
+    RCP<const Set> Interval::open() const {
+        return interval(start_, end_, true, true);
+    }
+
+    RCP<const Set> Interval::Lopen() const {
+        return interval(start_, end_, true, false);
+    }
+
+    RCP<const Set> Interval::Ropen() const {
+        return interval(start_, end_, false, true);
+    }
+
+    RCP<const Set> Interval::close() const {
+        return interval(start_, end_, false, false);
+    }
+
+    RCP<const Set> Interval::_intersection(const RCP<const Set> &o) const {
+        SYMENGINE_ASSERT(is_a<Interval>(*o));
+        const Interval &other = static_cast<const Interval &>(*o);
+        RCP<const Number> start, end;
+        bool left_open, right_open;
+        RCP<const Basic> start_end, end_start;
+        start_end = min({this->start_, other.end_});
+        end_start = min({this->end_, other.start_});
+
+        if (eq(*this->start_, *start_end) and eq(*other.start_ , *end_start)) {
+            RCP<const Basic> start_start, end_end;
+            start_start = min({this->start_, other.start_});
+            end_end = min({this->end_, other.end_});
+            if (neq(*this->start_, *other.start_)) {
+                if (eq(*this->start_, *start_start)) {
+                    start = other.start_;
+                    left_open = other.left_open_;
+                }
+                else {
+                    start = this->start_;
+                    left_open = this->left_open_;
+                }
+            }
+            else {
+                start = this->start_;
+                left_open = this->left_open_ or other.left_open_ ;
+            }
+
+            if (neq(*this->end_, *other.end_)) {
+                if (eq(*this->end_, *end_end)) {
+                    end = this->end_;
+                    right_open = this->right_open_;
+                }
+                else {
+                    end = other.end_;
+                    right_open = other.right_open_;
+                }
+            }
+            else {
+                end = this->end_;
+                right_open = this->right_open_ or other.right_open_;
+            }
+            return interval(start, end, left_open, right_open);
+        }
+        else {
+            return emptyset();
+        }
+    }
+
+    RCP<const Set> Interval::_union(const RCP<const Set> &o) const {
```

Comment:
Then use the C++ algorithm header notation. `set_union`, `set_intersection`, etc.